### PR TITLE
Get rid of `parse_type` from cairo-lang

### LINKED ISSUE #684: Replace usages of `starkware.cairo.lang.compiler.parser.parse_type` with own implementation
### Feature Request

It is used in the new data transformer.

### LINKED ISSUE #929: Implement `starkware.cairo.lang.compiler.ast.cairo_types`
### Feature Request

Used in `starknet_py.cairo.type_parser`

### PR DESCRIPTION
**Pull request by war-in**
Closes #929 
Closes #684 

## Introduced changes

- copy and minimize parser code from cairo-lang
- do not import anything new from cairo-lang

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: starknet_py/cairo/type_parser.py [hunk 1]
```diff
@@ -86,6 +85,7 @@ def _transform_cairo_lang_type(
                     OrderedDict(
                         (member.name, self._transform_cairo_lang_type(member.typ))
                         for member in cairo_type.members
+                        if member.name is not None  # just for typechecker
```
Comment (drknzz) [inline starknet_py/cairo/type_parser.py:88]: Hmm, right now the members will be filtered and a member without name just won't be included. Doesn't a member without a name mean that something went wrong? Would asserting that all names are present before the return be any better?

**Review by drknzz — APPROVED** [review]
🚀 🚀 🚀

**Review by cptartur — COMMENTED** [review]
Technically this can be breaking I think? Needs mention in migration guide at least. 

And I wonder if methods and classes from this module should be made internal?

File: docs/migration_guide.rst [hunk 2]
```diff
@@ -10,6 +10,14 @@ Breaking changes
 
 - :meth:`~starknet_py.hash.utils.pedersen_hash` function implementation has been changed, now uses only `crypto-cpp-py` library implementation.
 
+
+.. admonition:: Potentially breaking changes
+    :class: attention
+
+    Internal code of :meth:`starknet_py.abi.AbiParser.parse` has changed.
+    It should not affect users but keep in mind that the Contract can have difficulties resolving ABI.
+    If so please report.
```
Comment (cptartur) [inline docs/migration_guide.rst:19]: Maybe a link to gh issue form, but it's not necessary.

**Review by cptartur — APPROVED** [review]

**Pull request merged**